syntax = "proto3";

package collision_proto;

import "google/protobuf/empty.proto";

enum QueryType {
    HAS_VALUE = 0;
    EQUALS = 1;
    LESS_THAN = 2;
    GREATER_THAN = 3;
    CONTAINS = 4;
}

enum QueryFields {

    CRASH_DATE = 0;
    CRASH_TIME = 1;
    BOROUGH = 2 ;
    ZIP_CODE = 3;
    LATITUDE = 4;
    LONGITUDE = 5;
    LOCATION = 6;
    ON_STREET_NAME = 7;
    CROSS_STREET_NAME = 8 ;
    OFF_STREET_NAME = 9 ;
    NUMBER_OF_PERSONS_INJURED = 10 ;
    NUMBER_OF_PERSONS_KILLED = 11 ;
    NUMBER_OF_PEDESTRIANS_INJURED = 12 ;
    NUMBER_OF_PEDESTRIANS_KILLED = 13 ;
    NUMBER_OF_CYCLIST_INJURED = 14;
    NUMBER_OF_CYCLIST_KILLED = 15 ;
    NUMBER_OF_MOTORIST_INJURED = 16 ;
    NUMBER_OF_MOTORIST_KILLED = 17 ;
    CONTRIBUTING_FACTOR_VEHICLE_1 = 18 ;
    CONTRIBUTING_FACTOR_VEHICLE_2 = 19 ;
    CONTRIBUTING_FACTOR_VEHICLE_3 = 20 ;
    CONTRIBUTING_FACTOR_VEHICLE_4 = 21 ;
    CONTRIBUTING_FACTOR_VEHICLE_5 = 22 ;
    COLLISION_ID = 23 ;
    VEHICLE_TYPE_CODE_1 = 24 ;
    VEHICLE_TYPE_CODE_2 = 25 ;
    VEHICLE_TYPE_CODE_3 = 26 ;
    VEHICLE_TYPE_CODE_4 = 27 ;
    VEHICLE_TYPE_CODE_5 = 28 ;
    
}

message QueryCondition {
    QueryFields field = 1;
    QueryType type = 2;
    oneof data {

        string string_data = 3;
        uint32 uint8_data = 4;
        uint32 uint32_data = 5;
        uint64 uint64_data = 6;
        float float_data = 7;

    }

    optional bool not = 8;
    optional bool case_insensitive = 9;
}

message QueryRequest {
    optional uint64 id = 1;
    repeated uint32 requested_by = 2;
    repeated QueryCondition queries = 3;
}

message Collision {
    optional string crash_date = 1; 
    optional string crash_time = 2; 
    optional string borough = 3; 
    optional uint32 zip_code = 4; 
    optional float latitude = 5; 
    optional float longitude = 6; 
    optional string location = 7; 
    optional string on_street_name = 8; 
    optional string cross_street_name = 9; 
    optional string off_street_name = 10; 
    optional uint32 number_of_persons_injured = 11; 
    optional uint32 number_of_persons_killed = 12; 
    optional uint32 number_of_pedestrians_injured = 13; 
    optional uint32 number_of_pedestrians_killed = 14; 
    optional uint32 number_of_cyclist_injured = 15; 
    optional uint32 number_of_cyclist_killed = 16; 
    optional uint32 number_of_motorist_injured = 17; 
    optional uint32 number_of_motorist_killed = 18; 
    optional string contributing_factor_vehicle_1 = 19; 
    optional string contributing_factor_vehicle_2 = 20; 
    optional string contributing_factor_vehicle_3 = 21; 
    optional string contributing_factor_vehicle_4 = 22; 
    optional string contributing_factor_vehicle_5 = 23; 
    optional uint64 collision_id = 24; 
    optional string vehicle_type_code_1 = 25; 
    optional string vehicle_type_code_2 = 26; 
    optional string vehicle_type_code_3 = 27; 
    optional string vehicle_type_code_4 = 28; 
    optional string vehicle_type_code_5 = 29; 
}

message QueryResponse {
    uint64 id = 1;
    uint32 results_from = 2;
    repeated uint32 requested_by = 3;
    repeated Collision collision = 4;
}

service CollisionQueryService {
    rpc GetCollisions (QueryRequest) returns (QueryResponse);
    rpc StreamCollisions (QueryRequest) returns (stream QueryResponse);
    rpc SendRequest (QueryRequest) returns (google.protobuf.Empty);
    rpc ReceiveResponse (QueryResponse) returns (google.protobuf.Empty);
}
